stl/inc/semaphore

```diff
@@ -0,0 +1,85 @@
+// semaphore standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+
+#ifndef _SEMAPHORE_
+#define _SEMAPHORE_
+#include <yvals_core.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#if !_HAS_CXX20
+#pragma message("The contents of <semaphore> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <cstddef>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+template <ptrdiff_t _LeastMaxValue = 256>
+class counting_semaphore {
+public:
+    static constexpr ptrdiff_t max() noexcept {
+        // _LeastMaxValue is used here so that things such as binary_semaphore are
+        return _LeastMaxValue;
+    }
+    constexpr explicit counting_semaphore(ptrdiff_t _Desired) {
+        _STL_ASSERT(_Desired >= 0 && _Desired <= max(), "Desired value for the semaphore falls outside of semaphore range");
+        atomic_init(&_Counter, _Desired);
+    }
+    ~counting_semaphore()                         = default;
+    counting_semaphore(const counting_semaphore&) = delete;
+    counting_semaphore& operator=(const counting_semaphore&) = delete;
+
+    void release(ptrdiff_t _Update = 1) {
+        _STL_ASSERT(_Update >= 0 && _Update <= (max() - _Counter.load()), "The semaphore has been released more times than acquired");
+        _Counter.fetch_add(_Update, memory_order_release);
+    }
+    void acquire() {
+        while (!try_acquire()) {
+        }
+    }
+    bool try_acquire() noexcept {
+        ptrdiff_t _Old     = _Counter.load(memory_order_acquire);

```

Comment:
It would be better to assume the value of _Old rather than doing a load first.

Reply:
How would you assume the value of _Old? I've never heard this terminology before, and a cursory google search brings up no results in-context.

Reply:
Just pick something like 1. The first time through, the CAS is likely to fail; the CAS failure loads the actual value.

Reply:
I think CAS is too expensive load for semaphore in general, if it is almost certainly doomed to fail. So first attempt should be load.

"Just go ahead" strategy, however, would work well for binary semaphore, when there are only 1 or 0. It will actually just be an exchange, since you always leave it to zero.

Reply:
The standard just permits failing randomly on try_acquire IIRC so just using compare_exchange_weak like you did in your PR is fine, using compare_exchange_strong just strengthens the guarantee that it won't go negative AFAICT, it's been.... 6 months since I've touched this PR or thought about it really and the mindset that I had when I created it is therefore lost.